### PR TITLE
Handle flaky ACL tests by overriding assertDictEqual.

### PR DESCRIPTION
**Pull request by Lukasa**
This deals with the intermittent ACL manager unit test case failure that I've seen over the last few days by forcing unordered comparison of lists nested inside more complex dictionary structures.

### CODE REVIEW

File: calico/acl_manager/test/test_processor.py [hunk 1]
```diff
@@ -53,6 +54,22 @@ def tearDown(self):
                          'outbound': [],
                          'outbound_default': 'deny'}}
 
+    def assertDictEqual(self, d1, d2, msg=None):
+        """
+        This override for assertDictEqual ensures that lists do not need to
+        be ordered for dict comparison. It's an unfortunate hack, but c'est la
+        vie.
+        """
+        for k,v1 in d1.iteritems():
```
Comment (matthewdupre) [inline calico/acl_manager/test/test_processor.py]: This only tests if d1 is a subdictionary of d2.  Please add the reverse comparison (nicest to just loop this?).

File: calico/acl_manager/test/test_processor.py [hunk 2]
```diff
@@ -53,6 +54,22 @@ def tearDown(self):
...
+            self.assertIn(k, d2, msg)
+            v2 = d2[k]
+            if (isinstance(v1, collections.Iterable) and
+                not isinstance(v1, basestring)):
+                self.assertItemsEqual(v1, v2, msg)
+            else:
+                self.assertEqual(v1, v2, msg)
+        return True
```
Comment (matthewdupre) [inline calico/acl_manager/test/test_processor.py]: assertDictEqual shouldn't really return anything - it's probably harmless enough, but may as well remove so it matches the standard assertDictEqual.

**Comment by Lukasa** [comment]
@matthewdupre Done. =)

**Comment by Lukasa** [comment]
retest this please

Stupid Jenkins

**Comment by Lukasa** [comment]
retest this please

**Comment by Lukasa** [comment]
@plwhite These failures *are* Felix failures, which I have also seen intermittently fail. It'd be nice if you could take a look.

**Comment by Lukasa** [comment]
@plwhite Sadly, the problem seems to be an ordering one:

```
@@ -53,7 +53,7 @@

     -j felix-PREROUTING
   
   Chain felix-PREROUTING
-    -d 169.254.169.254/32 -p tcp -m tcp -j DNAT --dport 80 --to-destination 127.0.0.1:8775
+    -d 169.254.169.254/32 -p tcp -m tcp -j DNAT --to-destination 127.0.0.1:8775 --dport 80
   
 TABLE filter (IPv4)
   Chain FORWARD
@@ -110,6 +110,6 @@

     -j felix-PREROUTING
   
   Chain felix-PREROUTING
-    -d 169.254.169.254/32 -p tcp -m tcp -j DNAT --dport 80 --to-destination 127.0.0.1:8775
+    -d 169.254.169.254/32 -p tcp -m tcp -j DNAT --to-destination 127.0.0.1:8775 --dport 80
```

**Comment by Lukasa** [comment]
These failures are inevitable because `Rule.parameters` is a `dict`, and therefore has no defined ordering.

**Comment by Lukasa** [comment]
@plwhite I'm in a bit of a tough place here. It'd be nice to use `collections.OrderedDict`, but frankly the ordering is really only relevant to the tests, and otherwise we don't care. For that reason it feels stupid to bring in ordered dict and pay the price for maintaining that ordering in all cases.

It really just feels like that test is ill-considered: it cares about ordering in a situation where we really don't.

**Pull request merged**